kcp/virtual-workspace.sh

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env bash
+
+set -ex
+
+source ./kcp/utils.sh
+
+SERVICE_WS="service-$(echo $RANDOM)"
+USER_WS="user-$(echo $RANDOM)"
+
+export GITOPS_IN_KCP="true"
+
+cleanup_workspace() {
+    KUBECONFIG="${CPS_KUBECONFIG}" kubectl kcp ws
+    KUBECONFIG="${CPS_KUBECONFIG}" kubectl delete workspace $SERVICE_WS || true
+    KUBECONFIG="${CPS_KUBECONFIG}" kubectl delete workspace $USER_WS || true
+    pkill go
+}
+
+trap cleanup_workspace EXIT
+
+# Read the CPS KUBECONFIG path if it's not set already
+readKUBECONFIGPath
+
+echo "Initializing service provider workspace"
+createAndEnterWorkspace "$SERVICE_WS"
+
+echo "Creating APIExports and APIResourceSchemas in workspace $SERVICE_WS"
+KUBECONFIG="${CPS_KUBECONFIG}" make apply-kcp-api-all
+
+# Copy the identity hash from the backend APIExport in the service workspace so we can reference it later in the appstudio APIBinding
+identityHash=$(KUBECONFIG="${CPS_KUBECONFIG}" kubectl get apiexports.apis.kcp.dev gitopsrvc-backend-shared -o jsonpath='{.status.identityHash}')
+
+# Create permissions to bind APIExports. We need this workaround until KCP fixes the bug in their admission logic. Ref: https://github.com/kcp-dev/kcp/issues/1939  
+KUBECONFIG="${CPS_KUBECONFIG}" kubectl kcp ws
+bindingName=$(KUBECONFIG="${CPS_KUBECONFIG}" kubectl get clusterrolebinding | grep $SERVICE_WS | awk '{print $1}')
+userName=$(KUBECONFIG="${CPS_KUBECONFIG}" kubectl get clusterrolebindings $bindingName -o jsonpath='{.subjects[0].name}')
+
+KUBECONFIG="${CPS_KUBECONFIG}" kubectl kcp ws use $SERVICE_WS
+cat <<EOF | KUBECONFIG="${CPS_KUBECONFIG}" kubectl apply -f -
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: bind-apiexport
+rules:
+- apiGroups:
+  - apis.kcp.dev
+  resourceNames:
+  - gitopsrvc-backend-shared
+  - gitopsrvc-appstudio-shared
+  resources:
+  - apiexports
+  verbs:
+  - bind
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: bind-apiexport
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: bind-apiexport
+subjects:
+- apiGroup: rbac.authorization.k8s.io
+  kind: User
+  name: "$userName"
+EOF
+
+echo "Initializing user workspace"
+createAndEnterWorkspace "$USER_WS"
+
+createAPIBinding() {
```

Comment:
Similarly for this, recommend moving it to utils.sh^

Reply:
Same as above